pkg/controller/hibernation/azure_actuator.go

```diff
@@ -0,0 +1,231 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	"github.com/openshift/hive/pkg/azureclient"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+)
+
+const (
+	azurePowerStatePrefix  = "PowerState/"
+	azureUnknownPowerState = "unknown"
+)
+
+var (
+	azureRunningStates           = sets.NewString("running")
+	azureStoppedStates           = sets.NewString("stopped", "deallocated")
+	azurePendingStates           = sets.NewString("starting")
+	azureStoppingStates          = sets.NewString("stopping", "deallocating")
+	azureRunningOrPendingStates  = azureRunningStates.Union(azurePendingStates)
+	azureStoppedOrStoppingStates = azureStoppedStates.Union(azureStoppingStates)
+	azureNotRunningStates        = azureStoppedOrStoppingStates.Union(azurePendingStates)
+	azureNotStoppedStates        = azureRunningOrPendingStates.Union(azureStoppingStates)
+)
+
+func init() {
+	RegisterActuator(&azureActuator{azureClientFn: getAzureClient})
+}
+
+type azureActuator struct {
+	// azureClientFn is the function to build an Azure client, here for testing
+	azureClientFn func(*hivev1.ClusterDeployment, client.Client, log.FieldLogger) (azureclient.Client, error)
+}
+
+// CanHandle returns true if the actuator can handle a particular ClusterDeployment
+func (a *azureActuator) CanHandle(cd *hivev1.ClusterDeployment) bool {
+	return cd.Spec.Platform.Azure != nil
+}
+
+// StopMachines will stop machines belonging to the given ClusterDeployment
+func (a *azureActuator) StopMachines(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) error {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureRunningOrPendingStates, logger)
+	if err != nil {
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to stop")
+		return nil
+	}
+	errs := []error{}
+	for _, machineName := range azureMachineNames(machines) {
+		logger.WithField("machine", machineName).Info("Stopping cluster machine")
+		_, err = azureClient.DeallocateVirtualMachine(context.TODO(), clusterDeploymentResourceGroup(cd), machineName)
+		if err != nil {
+			errs = append(errs, err)
+			logger.WithError(err).WithField("machine", machineName).Error("Failed to stop machine")
+		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// StartMachines will select machines belonging to the given ClusterDeployment
+func (a *azureActuator) StartMachines(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) error {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureStoppedOrStoppingStates, logger)
+	if err != nil {
+		return err
+	}
+	if len(machines) == 0 {
+		logger.Warning("No machines were found to start")
```

Comment:
Same question here about whether this should be info level instead of warning. This would happen if there are machines in the Pending state.